python-primitiv/primitiv/_model.pyx

```diff
@@ -79,15 +39,41 @@ cdef class _Model:
 
     def add_parameter(self, str name, _Parameter param):
         self.wrapped.add_parameter(pystr_to_cppstr(name), param.wrapped[0])
+        self.added_parameters.append(param)
 
     def add_submodel(self, str name, _Model model):
         self.wrapped.add_submodel(pystr_to_cppstr(name), model.wrapped[0])
+        self.added_submodels.append(model)
+
+    def add_all_parameters(self):
+        for k, v in self.__dict__.items():
+            if isinstance(v, _Parameter) and v not in self.added_parameters:
+                self.add_parameter(k, v)
 
-    # NOTE(vbkaisetsu):
-    # get_parameter is replaced with `params` variable.
+    def add_all_submodels(self):
+        for k, v in self.__dict__.items():
+            if isinstance(v, _Model) and v not in self.added_submodels:
+                self.add_submodel(k, v)
 
-    # NOTE(vbkaisetsu):
-    # get_submodel is replaced with `submodels` variable.
+    def __getitem__(self, key):
+        cdef vector[string] names
+        if isinstance(key, str):
+            names.push_back(pystr_to_cppstr(key))
+        elif isinstance(key, tuple):
+            for name in key:
+                names.push_back(pystr_to_cppstr(name))
+        else:
+            raise TypeError("Argument 'key' has incorrect type (str or tuple)")
+        try:
+            return _Parameter.get_wrapper(&self.wrapped.get_parameter(names))
+        except:
+            try:
+                return _Model.get_wrapper(&self.wrapped.get_submodel(names))
+            except:
+                # NOTE(vbkaisetsu): DO NOT throw an exception here because
```

Comment:
If `get_parameter` function did not find a parameter with name `names`, the message from `get_parameter` would be killed and users may get an incorrect error message:  "**Model** with name `names` does not exist."

Reply:
No. Exceptions thrown by `get_submodel()` also be killed.
I fixed a comment and a typo, and added checking the message in the test case.
Does it work good?
```python
from primitiv import *
m = Model()
p = Parameter()
m.add_parameter("p", p)
sm = Model()
m.add_submodel("sm", sm)
m["p"]
# => <primitiv._parameter._Parameter at 0x7fdf04068870>
m["sm"]
# => <primitiv._model._Model at 0x7fdf040619f8>
m["xyz"]
# ---------------------------------------------------------------------------
# TypeError                                 Traceback (most recent call last)
# <ipython-input-9-86188247dc8a> in <module>()
# ----> 1 m["xyz"]
# 
# primitiv/_model.pyx in primitiv._model._Model.__getitem__()
# 
# TypeError: 'name' is not a name of neither parameter nor submodel

```